File: app/src/main/java/com/shenhua/zhidaodaily/utils/ThreadFactoryBuilder.java
package com.shenhua.zhidaodaily.utils;

import android.os.Process;
import android.support.annotation.NonNull;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

/**
 * Created by shenhua on 2017-11-16-0016.
 *
 * @author shenhua
 *         Email shenhuanet@126.com
 */
public class ThreadFactoryBuilder implements ThreadFactory {

    private String threadName = "demo-pool-";

    private ThreadFactoryBuilder() {
    }

    public ThreadFactoryBuilder setName(String name) {
        threadName = name;
        return this;
    }

    @Override
    public Thread newThread(@NonNull Runnable runnable) {
        int threadNum = 0;
        return new Thread(runnable, threadName + threadNum++) {
            @Override
            public void run() {
                Process.setThreadPriority(Process.THREAD_PRIORITY_BACKGROUND);
                super.run();
            }
        };
    }

    public static ExecutorService buildSimpleExecutorService() {
        ThreadFactory namedThreadFactory = Executors.defaultThreadFactory();
        return new ThreadPoolExecutor(1, 1,
                0L, TimeUnit.MILLISECONDS,
                new LinkedBlockingQueue<Runnable>(1024), namedThreadFactory, new ThreadPoolExecutor.AbortPolicy());
    }

}

File: app/src/main/java/com/shenhua/zhidaodaily/utils/ThreadFactoryBuilder.kt
package com.shenhua.zhidaodaily.utils

import android.os.Process
import java.util.concurrent.*

/**
 * Created by shenhua on 2017-11-16-0016.
 *
 * @author shenhua
 * Email shenhuanet@126.com
 */
class ThreadFactoryBuilder private constructor() : ThreadFactory {

    private var threadName = "demo-pool-"

    fun setName(name: String): ThreadFactoryBuilder {
        threadName = name
        return this
    }

    override fun newThread(runnable: Runnable): Thread {
        var threadNum = 0
        return object : Thread(runnable, threadName + threadNum++) {
            override fun run() {
                Process.setThreadPriority(Process.THREAD_PRIORITY_BACKGROUND)
                super.run()
            }
        }
    }

    companion object {

        fun buildSimpleExecutorService(): ExecutorService {
            val namedThreadFactory = Executors.defaultThreadFactory()
            return ThreadPoolExecutor(1, 1,
                    0L, TimeUnit.MILLISECONDS,
                    LinkedBlockingQueue(), namedThreadFactory, ThreadPoolExecutor.AbortPolicy())
        }

        fun buildFixedExecutorService(nThreads: Int): ExecutorService {
            val namedThreadFactory = Executors.defaultThreadFactory()
            return ThreadPoolExecutor(nThreads, nThreads,
                    0L, TimeUnit.MILLISECONDS,
                    LinkedBlockingQueue(), namedThreadFactory, ThreadPoolExecutor.AbortPolicy())
        }
    }

}
